packages/next/server/next-server.ts

```diff
@@ -1359,226 +1352,232 @@ export default class NextNodeServer extends BaseServer {
+              res.send()
+              return {
+                finished: true,
+              }
+            }
 
-          if (this.nextConfig.i18n) {
-            const localePathResult = normalizeLocalePath(
-              newUrl,
-              this.nextConfig.i18n.locales
-            )
-            if (localePathResult.detectedLocale) {
-              parsedDestination.query.__nextLocale =
-                localePathResult.detectedLocale
+            return {
+              finished: false,
             }
-          }
+          },
+        }
 
-          addRequestMeta(req, '_nextRewroteUrl', newUrl)
-          addRequestMeta(req, '_nextDidRewrite', newUrl !== req.url)
+        routes.push(middlewareCatchAllRoute)
+      }
+      if (this.getEdgeFunctions().length) {
+        const edgeCatchAllRoute: Route = {
+          match: getPathMatch('/:path*'),
+          type: 'route',
+          name: 'edge functions catchall',
+          fn: async (req, res, _params, parsed) => {
+            const edgeFunctions = this.getEdgeFunctions()
+            if (!edgeFunctions.length) return { finished: false }
+
+            const { query, pathname } = parsed
+            const normalizedPathname = removeTrailingSlash(pathname || '')
+            let page = normalizedPathname
+            let params: Params | undefined = undefined
+            let pageFound = !isDynamicRoute(page)
+
+            if (this.dynamicRoutes) {
+              for (const dynamicRoute of this.dynamicRoutes) {
+                params = dynamicRoute.match(normalizedPathname) || undefined
+                if (params) {
+                  page = dynamicRoute.page
+                  pageFound = true
+                  break
+                }
+              }
+            }
 
-          return {
-            finished: false,
-            pathname: newUrl,
-            query: parsedDestination.query,
-          }
-        }
+            if (!pageFound) {
+              return {
+                finished: false,
+              }
+            }
 
-        if (result.response.headers.has('x-middleware-refresh')) {
-          res.statusCode = result.response.status
-          for await (const chunk of result.response.body || ([] as any)) {
-            this.streamResponseChunk(res as NodeNextResponse, chunk)
-          }
-          res.send()
-          return {
-            finished: true,
-          }
-        }
+            const edgeSSRResult = await this.runEdgeFunction({
+              req,
+              res,
+              query,
+              params,
+              page,
+            })
 
-        return {
-          finished: false,
+            return {
+              finished: !!edgeSSRResult,
+            }
+          },
         }
-      },
-    }
 
-    const routes = []
-    if (!this.renderOpts.dev || devReady) {
-      if (this.getMiddleware()) routes[0] = middlewareCatchAllRoute
-      if (this.getEdgeFunctions().length) routes[1] = edgeCatchAllRoute
```

Comment:
This code looks weird because people might think that what if the first condition is false, but the second one is true?

Reply:
we always destruct these 2 and check them in the router matching, so the array length is always 2.